thorlcr/graph/thgraphslave.cpp

```diff
@@ -1872,22 +1837,67 @@ class CFileCache : public CInterface, implements IThorFileCache
     unsigned limit, purgeN;
     CriticalSection crit;
 
-    class CDelayedFileWapper : public CInterface, implements IDelayedFile
+    class CDelayedFileWapper : public CSimpleInterfaceOf<IDelayedFile>, implements IFileIO
     {
+        typedef CSimpleInterfaceOf<IDelayedFile> PARENT;
+
         CFileCache &cache;
+        CActivityBase &activity;
         Linked<CLazyFileIO> lFile;
+        CriticalSection crit;
+
     public:
-        IMPLEMENT_IINTERFACE;
+        IMPLEMENT_IINTERFACE_USING(PARENT);
 
-        CDelayedFileWapper(CFileCache &_cache, CLazyFileIO &_lFile) : cache(_cache), lFile(&_lFile) { }
+        CDelayedFileWapper(CFileCache &_cache, CActivityBase &_activity, CLazyFileIO &_lFile) : cache(_cache), activity(_activity), lFile(&_lFile) { }
 
         ~CDelayedFileWapper()
         {
             cache.remove(*lFile);
         }
         // IDelayedFile impl.
         virtual IMemoryMappedFile *queryMappedFile() { return lFile->queryMappedFile(); }
-        virtual IFileIO *queryFileIO() { return lFile->queryFileIO(); }
+        virtual IFileIO *queryFileIO() override { return this; }
+        virtual IFileIO *getFileIO() override { return lFile->getOpenFileIO(activity); }
```

Comment:
I assume you mean : "getFileIO returning something other than LINK(getFileIO). "

I did ponder whether to call the base method (in IDelayedFile) getOpenFileIO instead of getFileIO(). But I thought would look odd if base getOpenFileIO() return LINK(queryFileIO()).

An alternative clearer way perhaps is to get rid of queryFileIO() altogether, either by defining an alternative IDelayedFile variant (which is used widely) or by changing anywhere that uses queryFileIO() to use getFileIO()


Would that be clearer?